### PR TITLE
Add Flame (Ruby)

### PR DESCRIPTION
**Pull request by AlexWayfer**
Currently `rc` version.

### CODE REVIEW

**Comment by OvermindDL1** [comment]
I tested this with a couple other random servers that I already had built and got this, this does *not* look right, is flame really this slow?

╰─➤  tools/stats.exs -w 1 -d 3 _
Total Cores: 16
Concurrent Connections: 1000
Threads: 10
Warmup: 1 seconds
Duration: 3 seconds

Processing servers:
* bin/server_cpp_evhtp
* bin/server_crystal_router_cr
* bin/server_go_fasthttprouter
* bin/server_ruby_flame
* bin/server_ruby_rails
* bin/server_rust_iron
* bin/server_rust_nickel

Processing:  bin/server_cpp_evhtp
Processing:  bin/server_crystal_router_cr
Processing:  bin/server_go_fasthttprouter
Processing:  bin/server_ruby_flame
Processing:  bin/server_ruby_rails
Processing:  bin/server_rust_iron
Processing:  bin/server_rust_nickel


| Path | URL | Errors | Total Requests Count | Total Requests/s | Total Requests Throughput | Total Throughput/s | Req/s Avg | Req/s Stdev | Req/s Max | Req/s +/- | Latency Avg | Latency Stdev | Latency Max | Latency +/- | 50% | 75% | 90% | 99% |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| bin/server_cpp_evhtp | http://127.0.0.1:3000/ | 0 | 1924946 | 624847.35 | 117.49MB | 38.14MB | 64.06k | 15.96k | 108.68k | 70.67% | 2.30ms | 5.29ms | 216.69ms | 93.07% | 653.00us | 2.84ms | 6.26ms | 13.74ms |
| bin/server_cpp_evhtp | http://127.0.0.1:3000/user | 0 | 1860113 | 607361.23 | 113.53MB | 37.07MB | 61.90k | 15.68k | 103.54k | 77.33% | 2.61ms | 7.62ms | 217.55ms | 96.17% | 717.00us | 2.96ms | 6.74ms | 14.91ms |
| bin/server_cpp_evhtp | http://127.0.0.1:3000/user/0 | 0 | 1690966 | 550463.66 | 151.59MB | 49.35MB | 56.31k | 12.97k | 105.39k | 81.33% | 2.29ms | 5.05ms | 208.65ms | 94.03% | 840.00us | 2.92ms | 5.66ms | 12.97ms |
| bin/server_crystal_router_cr | http://127.0.0.1:3000/ | 0 | 208045 | 68582.04 | 12.30MB | 4.06MB | 6.97k | 375.88 | 8.01k | 65.33% | 14.39ms | 3.29ms | 31.70ms | 74.40% | 12.71ms | 17.99ms | 19.19ms | 25.49ms |
| bin/server_crystal_router_cr | http://127.0.0.1:3000/user | 0 | 184275 | 60973.94 | 10.90MB | 3.61MB | 6.17k | 641.74 | 7.34k | 77.33% | 15.77ms | 4.07ms | 220.76ms | 75.89% | 13.45ms | 19.68ms | 20.93ms | 27.22ms |
| bin/server_crystal_router_cr | http://127.0.0.1:3000/user/0 | 0 | 187307 | 61829.82 | 16.43MB | 5.42MB | 6.27k | 535.55 | 7.07k | 64.67% | 15.24ms | 3.78ms | 221.84ms | 84.12% | 13.65ms | 17.95ms | 20.81ms | 26.51ms |
| bin/server_go_fasthttprouter | http://127.0.0.1:3000/ | 0 | 1416603 | 457637.75 | 125.64MB | 40.59MB | 47.33k | 9.31k | 93.27k | 78.93% | 1.96ms | 1.83ms | 22.86ms | 87.75% | 1.30ms | 2.36ms | 4.19ms | 9.18ms |
| bin/server_go_fasthttprouter | http://127.0.0.1:3000/user | 0 | 1317451 | 430578.24 | 116.85MB | 38.19MB | 43.87k | 6.59k | 75.69k | 72.33% | 2.10ms | 1.97ms | 206.20ms | 89.95% | 1.70ms | 2.44ms | 4.01ms | 8.88ms |
| bin/server_go_fasthttprouter | http://127.0.0.1:3000/user/0 | 0 | 1275431 | 413076.2 | 199.48MB | 64.61MB | 42.47k | 7.86k | 72.81k | 74.33% | 2.24ms | 2.18ms | 42.37ms | 88.79% | 1.60ms | 2.66ms | 4.66ms | 10.87ms |
| bin/server_ruby_flame | http://127.0.0.1:3000/ | 0 | 8875 | 2917.49 | 329.35KB | 108.27KB | 330.02 | 140.30 | 1.01k | 73.70% | 5.40ms | 6.44ms | 54.79ms | 87.55% | 3.51ms | 7.57ms | 13.15ms | 30.22ms |
| bin/server_ruby_flame | http://127.0.0.1:3000/user | 0 | 7557 | 2487.93 | 280.44KB | 92.33KB | 316.12 | 141.93 | 1.03k | 73.75% | 6.33ms | 8.38ms | 89.83ms | 86.73% | 2.73ms | 9.15ms | 16.89ms | 38.00ms |
| bin/server_ruby_flame | http://127.0.0.1:3000/user/0 | 0 | 6388 | 2079.89 | 424.20KB | 138.12KB | 213.69 | 90.10 | 525.00 | 66.33% | 7.50ms | 8.60ms | 104.16ms | 87.15% | 5.07ms | 10.53ms | 18.23ms | 38.91ms |
| bin/server_ruby_rails | http://127.0.0.1:3000/ | 0 | 1918 | 631.33 | 327.78KB | 107.89KB | 128.33 | 9.37 | 141.00 | 72.00% | 7.81ms | 1.19ms | 30.41ms | 90.35% | 7.68ms | 8.18ms | 8.72ms | 10.37ms |
| bin/server_ruby_rails | http://127.0.0.1:3000/user | 0 | 1915 | 630.34 | 327.27KB | 107.72KB | 160.13 | 56.07 | 277.00 | 74.17% | 7.83ms | 1.01ms | 23.44ms | 82.92% | 7.78ms | 8.23ms | 8.73ms | 9.97ms |
| bin/server_ruby_rails | http://127.0.0.1:3000/user/0 | 0 | 1608 | 529.05 | 466.38KB | 153.45KB | 107.57 | 7.96 | 121.00 | 82.00% | 9.32ms | 1.20ms | 28.02ms | 87.00% | 9.28ms | 9.70ms | 10.23ms | 12.05ms |
| bin/server_rust_iron | http://127.0.0.1:3000/ | 0 | 1076468 | 349690.73 | 76.99MB | 25.01MB | 51.46k | 31.43k | 99.73k | 46.67% | 261.87us | 534.72us | 20.50ms | 98.06% | 198.00us | 275.00us | 423.00us | 1.13ms |
| bin/server_rust_iron | http://127.0.0.1:3000/user | 0 | 1028626 | 338904.95 | 73.57MB | 24.24MB | 49.23k | 28.38k | 100.38k | 61.43% | 271.08us | 485.81us | 33.75ms | 97.32% | 195.00us | 305.00us | 441.00us | 1.23ms |
| bin/server_rust_iron | http://127.0.0.1:3000/user/0 | 0 | 939644 | 307082.64 | 117.39MB | 38.36MB | 39.32k | 26.44k | 101.47k | 58.75% | 312.61us | 531.03us | 52.51ms | 95.82% | 214.00us | 343.00us | 552.00us | 1.82ms |
| bin/server_rust_nickel | http://127.0.0.1:3000/ | 0 | 985985 | 322166.47 | 122.24MB | 39.94MB | 110.26k | 10.01k | 144.77k | 68.89% | 47.67us | 23.06us | 5.76ms | 82.79% | 41.00us | 52.00us | 84.00us | 102.00us |
| bin/server_rust_nickel | http://127.0.0.1:3000/user | 0 | 760562 | 245334.51 | 94.29MB | 30.42MB | 123.36k | 13.53k | 139.37k | 70.97% | 59.75us | 40.45us | 11.81ms | 99.22% | 61.00us | 77.00us | 85.00us | 99.00us |
| bin/server_rust_nickel | http://127.0.0.1:3000/user/0 | 0 | 781063 | 257566.94 | 122.91MB | 40.53MB | 130.89k | 4.40k | 139.48k | 63.33% | 57.72us | 18.08us | 1.29ms | 60.75% | 60.00us | 71.00us | 79.00us | 93.00us |


# Rankings

## Ranking by Average Requests per second:
1. 594224 req/sec : bin/server_cpp_evhtp
2. 433764 req/sec : bin/server_go_fasthttprouter
3. 331892 req/sec : bin/server_rust_iron
4. 275022 req/sec : bin/server_rust_nickel
5. 63795 req/sec : bin/server_crystal_router_cr
6. 2495 req/sec : bin/server_ruby_flame
7. 596 req/sec : bin/server_ruby_rails

**Comment by AlexWayfer** [comment]
Answered in the issue, where this comment appeared early.

File: ruby/flame/application.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+module FlameTest
+  # Application that will be launched
+  class Application < Flame::Application
+    mount :index, '/'
+    mount :user
+
+    p router.routes
```
Comment (AlexWayfer) [inline ruby/flame/application.rb]: Need to delete this. :grimacing:

**Comment by AlexWayfer** [comment]
Need be changed after merging #193.

**Comment by waghanza** [comment]
@AlexWayfer could you update this ?

**Comment by AlexWayfer** [comment]
> @AlexWayfer could you update this ?

Yes, sure. Updated. Sorry for delay.

**Comment by AlexWayfer** [comment]
My results (seems working):

```
> sudo bin/benchmarker ruby
Last update: 2018-05-08

OS: Linux (version: 4.16.5-1-ARCH, arch: x86_64)
CPU Cores: 4
threads: 5, requests: 1000.0

Benchmark running ...
Done. <- rails
Done. <- sinatra
Done. <- roda
Done. <- rack-routing
Done. <- flame

### Ranking (Framework)

1. [roda](https://github.com/jeremyevans/roda) (ruby)
2. [rack-routing](https://github.com/iAmPlus/rack-routing) (ruby)
3. [flame](https://github.com/AlexWayfer/flame) (ruby)
4. [sinatra](https://github.com/sinatra/sinatra) (ruby)
5. [rails](https://github.com/rails/rails) (ruby)

### Ranking (Language)

1. ruby ([roda](https://github.com/jeremyevans/roda))

### All frameworks

| Language (Runtime)        | Framework (Middleware)    |       Max [sec] |       Min [sec] |       Ave [sec] |
|---------------------------|---------------------------|-----------------|-----------------|-----------------|
| ruby                      | rails                     |        9.674696 |        6.494990 |        8.104362 |
| ruby                      | sinatra                   |        2.890795 |        1.503194 |        2.107554 |
| ruby                      | roda                      |        0.916863 |        0.593624 |        0.729372 |
| ruby                      | rack-routing              |        1.167781 |        0.741380 |        1.001433 |
| ruby                      | flame                     |        1.932608 |        1.440200 |        1.653960 |
```

**Pull request merged**